### ISSUE TITLE
[Question] How should I interpretate evaluate_acquisition_function() = 0.0?

### ISSUE THREAD

**Issue by amgonzalezd** [facebook/Ax#1445]
Hey guys!
I'm trying to track the values of the acquisition function at each trial of my experiment. I'm primarly using analytic acquisistion functions such as EI, noisyEI, PI and UCB.
I'm using the Service API with BOTORCH_MODULAR (and 5 Sobol steps) and calling model_bridge.evaluate_acquisition_function like this
```
acquisition_score = []
for i in range(10):
    parameters, trial_index = ax_client.get_next_trial()
   
    model_bridge = ax_client.generation_strategy.model
    obs_feature = ObservationFeature(parameters=parameters)

    acq_score = model_bridge.evaluate_acquisition_function([obs_feat])
    acquisition_score.append(acq_score)

    ax_client.complete_trial(trial_index=trial_index, raw_data=evaluation_function(parameters))
```

Sometimes (but not always) I get as a result acq_score = 0.0.

I've analyze this a bit and it happens in three diferent scenarios:
- At the begining of the botorch model (after the 5 sobol trials) during trial 6 or 7
- When arms are being repeated --> I understand that in that case the acq_function is returning 0... but why is then my model suggesting the same parameters, that I've already seen?
- At trials where I get the following warning:  `NumericalWarning: Negative variance values detected. This is likely due to numerical instabilities. Rounding negative variances up to 1e-10.`

Thanks in advance!

**Comment by mpolson64** [comment]
When this happens it is because Ax is confident that no point would improve on the best point observed. A value of zero here is indicating the the expected improvement or probability of improvement is nothing. Assuming your model fit is good (which you can check via cross validation) this typically means it is safe to stop the optimization. If the model fit is bad, you could consider using more Sobol points or using a different surrogate, like our SaasFullyBayesianSingleTaskGP which is able to get much better fits for high dimensional search spaces.

Let me know if this was helpful!

**Comment by Balandat** [comment]
> A value of zero here is indicating the the expected improvement or probability of improvement is nothing

Mathematically, it should never happen that either of these are zero, but it can happen numerically. @SebastianAment has done a bunch of work recently on [logEI](https://github.com/pytorch/botorch/blob/main/botorch/acquisition/analytic.py#L367-L383) (and various other log variants of other acquisition functions) that is much less prone to these numerical issues, and your application could be a nice test case for it.

I definitely second Miles' point about checking the model fit though.

**Comment by amgonzalezd** [comment]
Thanks for the quick replies!!
I've  tracked the model fit with cross validation :) The table below shows the results of one example. 
The ACQ column results from `evaluate_model_fit()` at each trial and the R2 column is the R2 score using the results of the cross validation (with a single item test). 
The metric is a branin synthetic problem adapted by Forrester 2008 (Engineering Design via Surrogate Modelling).
I'm using a SingleTaskGP with a Matern 0.5 kernel and ARD and EI as acquisition function. 

![image](https://user-images.githubusercontent.com/69913576/221012256-3e82cddb-51ec-44cd-a763-f66e3929729d.png)



In trial 6 the value of the acquisition function is 0.0 and the model fit (R2) is -3e-14.
After that it finds a better point in trial 7, raising the r2 score to 0.29 and on trial 8 the acquisition score is once again 0.000.
Afterwards the model is just getting better.

I've also tried other Kernels (RBF, Matern 1.5, Matern 2.5) and other acquisition functions (PI, UCB, noisyEI) and in some combinations the same problem occurrs.

I will definitely try logEI, but I would kinda like to understand why is this happening :)

Thanks a lot!!

**Comment by lena-kashtelyan** [comment]
@Balandat, @SebastianAment, @saitcakmak perhaps you could chime in here?

**Comment by Balandat** [comment]
It's kind of hard to say without seeing the logs or the specific example. There are a number of things that could cause this. The most basic explanation is that the new data point that was acquired happened to be such that the model wasn't able to make much sense of it (hence the poor R2). What exactly happened would require inspecting the model. The good news is that even if that happens, this will usually result in the acquisition function degenerating to random sampling in that iteration, which will often improve the model and enable more principled optimization down the road (this seems to be what's happening in your case).

**Comment by lena-kashtelyan** [comment]
cc @amgonzalezd, let us know if the answer above helps or whether you still needed assistance!

**Comment by SebastianAment** [comment]
Hi @amgonzalezd,

an R2 of `10e-14` implies that the model is predicting the mean of the observed target values, virtually to machine precision. For a stationary Gaussian process, this can happen due to two primary reasons, which can also cause the acquisition values to vanish:

1) The length scale(s) of the kernel are very small, causing the mean predictions to revert back to the prior mean quickly when moving away from observed data points. With the output standardization that Ax automatically applies, the prior mean coincide with the empirical mean of the data, hence the R2 of zero. If the prior mean is much worse than the best observed value the expected improvement value is close or numerically exactly zero, which makes acquisition function optimization less likely to be successful.
2) The output scale of the kernel (if you're using a `ScaleKernel` wrapper) is exceedingly close to zero, which makes the entire process close to constant everywhere.

Printing out the kernel hyper-parameters during your optimization run can determine this for certain.

While this isn't the primary culprit here, just wanted to point out that Matern 1/2 gives rise to *very* rough sample paths, and is non-differentiable almost everywhere. A higher degree of smoothness (nu => 2.5) can be assumed for most applications.

**Comment by lena-kashtelyan** [comment]
Hi @amgonzalezd, I'm closing this issue as inactive. I hope your questions were resolved with @SebastianAment's most recent follow-up! Please reopen the issue if you follow up on it, as we are unlikely to see new comments on closed issues.
